### PR TITLE
Lindsay + Katricia - Edges - Adagrams

### PR DESCRIPTION
**Pull request by krsmith7**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | A method signature, parameters, method body with implicit or explicit return, and end. 
What are the advantages of using git when collaboratively working on one code base? | Logs of all changes by all users, ability to edit at the same time, ability to merge changes and manage conflicts.
What kind of relationship did you and your pair have with the unit tests? | They provided us guidance on next steps and ways to fix our code. 
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used enumerables in transforming the array and hash data more efficiently.
What was one method you and your pair used to debug code? | We used pry. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We discussed ways to step outside of "guess culture" tendencies when giving feedback. We also discussed if the driver-navigator framework worked for us.

### CODE REVIEW

File: adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,136 @@
+require 'pry'
+
```
Comment (droberts-sea) [inline adagrams.rb:2]: You've included this file twice, once as `adagrams.rb`, and once as `lib/adagrams.rb`.

File: specs/adagrams_spec.rb [hunk 2]
```diff
@@ -10,12 +10,15 @@
 describe 'Adagrams' do
   describe 'draw_letters method' do
     it 'draws ten letters from the letter pool' do
-      drawn_letters = draw_letters
+      #Arrange
+      drawn_letters = ["A", "Q", "L", "J", "P", "E", "K", "I", "A", "Z"]
+      # Assert
```
Comment (droberts-sea) [inline specs/adagrams_spec.rb:15]: I'm curious why you made this change. This test is supposed to be testing the `draw_letters` method, but with this change that method is never called in this test. All you do now is create an array by hand and then check that it has 10 letters.

File: specs/adagrams_spec.rb [hunk 3]
```diff
@@ -56,116 +59,118 @@
...
-      expect(score_word(words.first)).must_equal 18
-      expect(score_word(words.last)).must_equal 18
-
-      best_word = highest_score_from words
+#   end
+#
+#   describe 'highest_score_from method' do
+#     it 'returns a hash that contains the word and score of best word in an array' do
+#       words = ['X', 'XX', 'XXX', 'XXXX']
+#       best_word = highest_score_from words
```
Comment (droberts-sea) [inline specs/adagrams_spec.rb:90]: Why are these tests commented out? Getting these to run is part of the assignment.

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+#create method to see if word uses letters that are actually available
+def uses_available_letters? (input, drawn_letters)
+  is_valid = "default"
+  word_letters = make_word_array(input)
+  word_letters.each do |letter|
+    if drawn_letters.include?letter
+      is_valid = true
+      puts "letter valid " + "#{is_valid}"
+      letter_index = drawn_letters.index(letter)
```
Comment (droberts-sea) [inline lib/adagrams.rb:43]: Please remove debugging output before submitting projects.

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  end
+  score += 8 if word.length >= 7 && word.length <= 10
+  return score
+end
+
+
+#create a method to return a hash with the highest scoring words
+def highest_score_from(words)
+  all_words_scores = []
+
```
Comment (droberts-sea) [inline lib/adagrams.rb:103]: This implementation is close, but the tie-breaking logic isn't quite right. The problem is that you don't respect the order in which the words were given.

The reason this happens is the sorting you do on line 112. This will re-order the words by their scores, so that the input order is now gone.

As it turns out, you don't need to sort the words at all to find the max. If we change line 112 to `word_rank = all_words_scores`, the rest of your logic works perfectly and all the (uncommented) tests pass.

File: lib/adagrams.rb [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    if word_rank[index][:score] > best_word[:score]
+      best_word = word_rank[index]
+
+    # Tie Handling
+    elsif word_rank[index][:score] == best_word[:score]
+      if best_word[:word].length != 10
+        if word_rank[index][:word].length == 10
+          best_word = word_rank[index]
+        elsif word_rank[index][:word].length < best_word[:word].length
```
Comment (droberts-sea) [inline lib/adagrams.rb:128]: Good work getting this tricky tie-breaking logic figured out.

File: lib/adagrams.rb [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+
+#turn all_letters array into a hash that represents the score of each letter
+LETTER_SCORES = {}
+
+# Make hash for letter to score
+all_letters.uniq.each do |letter|
+
+  case letter
+  when "A","E", "I", "O", "U", "L", "N", "R", "S", "T"
```
Comment (droberts-sea) [inline lib/adagrams.rb:65]: The idea of generating this data structure dynamically with a loop and a case statement is an interesting one. However, I think I would argue that the resulting code is a little less readable than just hardcoding it:

```ruby
LETTER_SCORES = {
  "A" => 1,
  "B" => 3,
  # ...
}
```

File: lib/adagrams.rb [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+end
+
+
+# Define method for creating user hand
+def draw_letters(array)
+  user_letters = array.sample(10)
+  return user_letters
+end
+
+DRAWN_LETTERS = draw_letters(all_letters)
```
Comment (droberts-sea) [inline lib/adagrams.rb:25]: I think there was some confusion about what was expected here. The idea is that `draw_letters` should be called by the driver code (`wave-X-game.rb`) or by the specs in order to create a hand of letters. Your code should not ever need to call `draw_letters`.

File: lib/adagrams.rb [hunk 9]
```diff
@@ -0,0 +1,136 @@
+require 'pry'
+
+#create data structure
+all_letters = []
+
+letters = {"A" => 9, "B" => 2, "C" => 2, "D"=> 4, "E" => 12, "F"=> 2, "G"=>3, "H"=>2, "I" => 9, "J"=>1, "K"=> 1, "L"=>4, "M"=> 2, "N"=> 6, "O"=>8, "P" => 2, "Q" => 1, "R"=>6, "S"=>4, "T"=>6, "U"=>4, "V"=>2, "W"=>2, "X"=>1, "Y"=>2, "Z"=>1}
+
+
+# Fill pool of letters array with letters
+letters.each do |letter, freq|
```
Comment (droberts-sea) [inline lib/adagrams.rb:10]: I like the way you've chosen to store the letter frequencies in a hash and then dynamically turn them into an array. I would say that this is much more readable than a giant hard-coded array with 9 As, 2 Bs, 2 Cs, etc.

File: lib/adagrams.rb [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  input = input.upcase.split("")
+  return input
+end
+
+
+#create method to see if word uses letters that are actually available
+def uses_available_letters? (input, drawn_letters)
+  is_valid = "default"
+  word_letters = make_word_array(input)
+  word_letters.each do |letter|
```
Comment (droberts-sea) [inline lib/adagrams.rb:39]: This method is destructive! Because you call `delete_at` on the `drawn_letters` directly, the hand will be reduced by the letters in the input. For example:

```ruby
hand = draw_letters
puts hand
# => ["T", "T", "E", "F", "R", "A", "E", "L", "V", "A"]
uses_available_letters?('tte', hand)
puts hand
# => ["F", "R", "A", "E", "L", "V", "A"]
```

While none of the tests we provided check for destroying the hand, this is bad because it's unexpected behavior. Nothing about "check whether this word is made up of letters in this hand" suggests that the hand will be changed in the process, but that's exactly what happens.

The way to address this problem would be to make a copy of the hand before removing letters from it.

**Comment by droberts-sea** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Both teammates contributed to the codebase | yes
Small commits with meaningful commit messages | some - I would like to see both more frequent commits, and more descriptive commit messages. Each message should give the reader a sense of what changed and why, without having to dive into the code or the project specs.
**Code Requirements** |
`draw_letters` method |
Uses appropriate data structure to store the letter distribution | yes
All tests for `draw_letters` pass | no - You changed the tests! It seems like there was a requirements miss here - see inline.
`uses_available_letters?` method |
All tests for `uses_available_letters?` pass | yes
`score_word` method |
Uses appropriate data structure to store the letter scores | sort of - see inline
All tests for `score_word` pass | yes
`highest_score_from` method |
Appropriately handles edge cases for tie-breaking logic | almost - see inline
All tests for `highest_score_from` pass | almost - see inline
**Overall** |

I ended up having a lot to say about this one. In general your product code (`lib/adagrams.rb`) does a good job of solving the problem. There are a few places where it is incorrect or could be cleaned up, but in general I feel pretty good about it.

I am more concerned by what I see as trouble with the process. In particular, it seems like you struggled with the idea of writing code to match the tests. The tests (and in this case the driver code) should give you an idea of what your code should look like from the outside, how it should be used. Being able to read the tests and quickly turn this into a high-level code structure is a skill that will continue to be important as we move through the curriculum. You should be running tests on a regular basis, and if you must comment some out, uncomment them an make sure they pass before submitting.

In general I think you are on a good trajectory. Going forward, it will be important to make sure you have a strong understanding of what is expected and how you are supposed to write your code. This will be particularly relevant on Grocery Store, which has a similar structure to Adagrams.

Let me know if you have any questions about that, and keep up the hard work!
